### PR TITLE
os/bluestore: add cache-related stats

### PR DESCRIPTION
**Pull request by xiexingguo**

### CODE REVIEW

**Comment by xiexingguo** [comment]
Passed local test:
```
==> rm -r store_test_temp_dir
2016-09-02 18:50:17.501847 7fed14609640  1 bluestore(store_test_temp_dir) fsck finish with 0 errors
[       OK ] ObjectStore/StoreTest.TooManyBlobsTest/2 (22748 ms)
[----------] 60 tests from ObjectStore/StoreTest (1815983 ms total)

[----------] Global test environment tear-down
[==========] 60 tests from 1 test case ran. (1815983 ms total)
[  PASSED  ] 60 tests.

**Comment by xiexingguo** [comment]
retest this please

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -569,10 +573,13 @@ class BlueStore : public ObjectStore,
 
   /// a cache (shard) of onodes and buffers
   struct Cache {
+    BlueStore *store;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: Seems you need that for accessing BlueStore::logger. Perhaps it's better to have direct pointer to logger here instead.

File: src/os/bluestore/BlueStore.h [hunk 2]
```diff
@@ -569,10 +573,13 @@ class BlueStore : public ObjectStore,
...
     std::mutex lock;                ///< protect lru and other structures
 
     static Cache *create(string type);
 
+    void set_store(BlueStore *bs) { store = bs; }
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: Shouldn't we pass this pointer (or logger one) via create method?

**Comment by ifed01** [comment]
LGTM

File: src/os/bluestore/BlueStore.h (hunk 2 above)
Comment (xiexingguo) [inline src/os/bluestore/BlueStore.h]: @ifed01 Thanks, Igor.
 I tried but got various compiling errors, such as:
```
In file included from /home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:22:0:
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.h: In member function ‘void BlueStore::BufferSpace::read(uint64_t, uint64_t, BlueStore::ready_regions_t&, interval_set<long unsigned int>&)’:
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.h:1211:17: error: invalid use of non-static data member ‘BlueStore::logger’
   PerfCounters *logger;
                 ^
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:1029:3: error: from this location
   logger->inc(l_bluestore_buffer_hit_bytes, hit_bytes);

```

Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: @xiexingguo Probably I missed something but my suggestion was just to modify create in the following manner:

static Cache *create(string type, PerfCounters* log)
cache = new ...
cache->logger = log
return cache;
}
I don't expect any issues with such code...

Comment (xiexingguo) [inline src/os/bluestore/BlueStore.h]: @ifed01 
I fully understand what you mean. I combined the above two comments you suggested into local branch but made compiler unhappy.

The code(partial):

```diff

-BlueStore::Cache *BlueStore::Cache::create(string type)
+BlueStore::Cache *BlueStore::Cache::create(string type, PerfCounters *logger)
 {
+  Cache *c;
   if (type == "lru")
-    return new LRUCache;
-  if (type == "2q")
-    return new TwoQCache;
-  assert(0 == "unrecognized cache type");
+    c = new LRUCache;
+  else if (type == "2q")
+    c = new TwoQCache;
+  else
+    assert(0 == "unrecognized cache type");
+  c->logger = logger;
+  return c;
 }

@@ -1075,10 +1087,12 @@ BlueStore::OnodeRef BlueStore::OnodeSpace::lookup(const ghobject_t& oid)
   ceph::unordered_map<ghobject_t,OnodeRef>::iterator p = onode_map.find(oid);
   if (p == onode_map.end()) {
     dout(30) << __func__ << " " << oid << " miss" << dendl;
+    logger->inc(l_bluestore_onode_misses);
     return OnodeRef();
   }
   dout(30) << __func__ << " " << oid << " hit " << p->second << dendl;
   cache->_touch_onode(p->second);
+  logger->inc(l_bluestore_onode_hits);
   return p->second;
 }

@@ -2767,7 +2814,7 @@ void BlueStore::set_cache_shards(unsigned num)
   assert(num >= old);
   cache_shards.resize(num);
   for (unsigned i = old; i < num; ++i) {
-    cache_shards[i] = Cache::create(g_conf->bluestore_cache_type);
+    cache_shards[i] = Cache::create(g_conf->bluestore_cache_type, logger);
   }
 }

@@ -569,9 +573,10 @@ public:
 
   /// a cache (shard) of onodes and buffers
   struct Cache {
+    PerfCounters *logger;
     std::mutex lock;                ///< protect lru and other structures
 
-    static Cache *create(string type);
+    static Cache *create(string type, PerfCounters *);
 
     virtual ~Cache() {}

```

It keeps complaining:
```
n file included from /home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:22:0:
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.h: In member function ‘BlueStore::OnodeRef BlueStore::OnodeSpace::lookup(const ghobject_t&)’:
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.h:1209:17: error: invalid use of non-static data member ‘BlueStore::logger’
   PerfCounters *logger;
                 ^
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:1090:5: error: from this location
     logger->inc(l_bluestore_onode_misses);
     ^
In file included from /home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:22:0:
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.h:1209:17: error: invalid use of non-static data member ‘BlueStore::logger’
   PerfCounters *logger;
                 ^
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:1095:3: error: from this location
   logger->inc(l_bluestore_onode_hits);
   ^
```

So we can leave it this way... the difference is slight...

Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: IMHO this 
+    logger->inc(l_bluestore_onode_misses);
just needs to be replaced with 
+    cache->logger->inc(l_bluestore_onode_misses);
The same at the second location..

**Comment by liewegas** [comment]
:+1:

File: src/os/bluestore/BlueStore.h (hunk 2 above)
Comment (xiexingguo) [inline src/os/bluestore/BlueStore.h]: @ifed01 Yeah, that works!
Thanks, Igor!

**Pull request merged**